tcptracer-bpf.c

```diff
@@ -854,42 +939,54 @@ static int tcp_send(struct pt_regs *ctx, const size_t size) {
-				if (check_family(sk, status, AF_INET)) {
-					if (read_ipv4_tuple(&t, status, sk)) {
-						t.lport = ntohs(t.lport); // Making ports human-readable
-						t.rport = ntohs(t.rport);
-
-						struct tracked_socket *res = bpf_map_lookup_elem(&tracked_sockets, &sk);
-						u64 current_time = bpf_ktime_get_ns();
-						u64 ttfb = 0;
-						struct tracked_socket tsocket = {.active = 1};
-						tsocket.prev_send_time_ns = current_time;
-						bpf_map_update_elem(&tracked_sockets, &sk, &tsocket, BPF_ANY);
-						if (res != NULL) {
-							ttfb = current_time - res->prev_send_time_ns;
-						}
-						u64 cpu = bpf_get_smp_processor_id();
-						int http_status_code = 0;
-						u16 mysql_greeting_protocol_version = 0;
-						if (parse_http_response(data, iov.iov_len, &http_status_code)) {
-							send_http_response(ctx, t, http_status_code, ttfb / 1000, current_time, cpu);
-						} else if (parse_mysql_greeting(data, iov.iov_len, &mysql_greeting_protocol_version)) {
-							send_mysql_greeting(ctx, t, mysql_greeting_protocol_version, current_time, cpu);
-						}
-					}
-				}
+	struct msghdr msg = {};
+	bpf_probe_read(&msg, sizeof(msg), k_msg);
+//	bpf_debug("test %d <> %d\n", msg.msg_iter.type & ~(READ | WRITE), status->iter_type);
+
+	if ((msg.msg_iter.type & ~(READ | WRITE)) == status->iter_type) {
+		char *data = bpf_map_lookup_elem(&write_buffer_heap, &zero);
+		if (data != NULL) {
+			struct iovec iov = {};
+			bpf_probe_read(&iov, sizeof(iov), (void *)msg.msg_iter.iov);
+			bpf_probe_read(data, MAX_MSG_SIZE, iov.iov_base);
+
+			struct ipv4_tuple_t t = {};
+			struct ipv6_tuple_t t6 = {};
+
+			if(!is_v4_or_v6(sk, status)){
+			    return 0;
 			}
+
+			struct tracked_socket *res = bpf_map_lookup_elem(&tracked_sockets, &sk);
+            u64 current_time = bpf_ktime_get_ns();
+            u64 ttfb = 0;
+            struct tracked_socket tsocket = {.active = 1};
+            tsocket.prev_send_time_ns = current_time;
+            bpf_map_update_elem(&tracked_sockets, &sk, &tsocket, BPF_ANY);
+            if (res != NULL) {
+                ttfb = current_time - res->prev_send_time_ns;
+            }
+            u64 cpu = bpf_get_smp_processor_id();
+            int http_status_code = 0;
+            u16 mysql_greeting_protocol_version = 0;
+
+            if(is_ipv4(sk, status)) {
+                get_ip_v4_tuple(&t, sk, status);
+
+                if (parse_http_response(data, iov.iov_len, &http_status_code)) {
+                    send_http_response(ctx, t, http_status_code, ttfb / 1000, current_time, cpu);
+                } else {
+                    send_mysql_greeting(ctx, t, mysql_greeting_protocol_version, current_time, cpu);
+                }
+            } else {
+                get_ip_v6_tuple(&t6, sk, status);
+
+                if (parse_http_response(data, iov.iov_len, &http_status_code)) {
+                    send_http_response_v6(ctx, t6, http_status_code, ttfb / 1000, current_time, cpu);
+                } else {
```

Comment:
`else if (parse_mysql_greeting(data, iov.iov_len, &mysql_greeting_protocol_version))` was removed only for debugging purposes, it should be here, otherwise we gonna mark every other connection as mysql! :D

Reply:
Handled it (this is probably part of the debug you were doing before you and @julianosk jumped on that other ticket, so I missed it :frowning_face: )

Reply:
Put it back